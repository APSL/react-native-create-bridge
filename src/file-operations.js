const path = require("path");
const fs = require("mz/fs");

exports.pkg = require(path.join(process.cwd(), "package.json"));

function getFileNames(dirPath) {
  return fs.readdir(dirPath).catch(e => console.error("[getFileNames] ", e));
}

function readFile(file, readDirPath) {
  return fs
    .readFile(path.join(readDirPath, file), "utf-8")
    .catch(e => console.error("[readFile] ", e));
}

function parseFile(fileData, templateName, packageName = null, app = null) {
  return fileData
    .replace(/{{template}}/g, templateName)
    .replace(/{{packageName}}/g, packageName)
    .replace(/{{app}}/g, app);
}

function readAndWriteFiles(
  files,
  paths,
  templateName,
  packageName = null,
  app = null
) {
  const { readDirPath, writeDirPath } = paths;
  return Promise.all(
    files.map(file => {
      readFile(file, readDirPath).then(fileData => {
        const parsedFile = parseFile(fileData, templateName, packageName, app);
        const fileName = file.replace("Template", templateName);
        return fs.writeFile(path.join(writeDirPath, fileName), parsedFile);
      });
    })
  ).catch(e => console.error("[readAndWriteFiles] ", e));
}

module.exports = {
  getFileNames,
  readFile,
  parseFile,
  readAndWriteFiles
};
