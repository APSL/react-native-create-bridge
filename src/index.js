#! /usr/bin/env node

import inquirer from 'inquirer'
import path from 'path'
import fs from 'mz/fs'
import isValid from 'is-valid-path'
import mkdir from 'mkdirp-promise'

const templateNameRegex = /\w+/
const pkg = require(path.join(process.cwd(), 'package.json'))
const promptConfig = [
  {
    type: 'input',
    name: 'templateName',
    message: 'What is your bridge module called?',
    default: 'ExampleBridge',
    validate: (input) => templateNameRegex.test(input),
  },
  {
    type: 'checkbox',
    name: 'environment',
    message: 'What OS & languages would you like to support?',
    default: ['Android/Java', 'iOS/Swift', 'iOS/Objective-C'],
    choices: ['Android/Java', 'iOS/Swift', 'iOS/Objective-C']
  },
  {
    type: 'input',
    name: 'jsPath',
    message: 'What directory should we deliver your JS files to?',
    default: '.',
    validate: (input) => isValid(input),
  }
]

const environmentMap = {
  'Android/Java': createAndroidEnvironment,
  'iOS/Swift': createSwiftEnvironment,
  'iOS/Objective-C': createObjCEnvironment
}

export function isRNVersionNew ({ dependencies }) {
  return parseFloat(dependencies["react-native"]) >= 0.4
}

async function init () {
  try {
    const answers = await inquirer.prompt(promptConfig)

    const promises = answers.environment.map(async function(env) {
      return environmentMap[env](answers.templateName)
    })

    promises.push(createJSEnvironment(answers.templateName, answers.jsPath))

    await Promise.all(promises)

    console.log('Your bridge was successfully created! ðŸŽ‰')
  } catch (e) {
    console.log(e)
  }
}

export function parseFile (fileData, templateName, packageName = null, app = null) {
  const isRNNew = isRNVersionNew(pkg)

  const iOSHeader = isRNNew ? '<React/' : '"'
  const iOSCloser = isRNNew ? '>' : '"'

  return fileData.replace(/{{template}}/g, templateName)
    .replace(/{{packageName}}/g, packageName)
    .replace(/{{app}}/g, app)
    .replace(/{{iOSHeader}}/g, iOSHeader)
    .replace(/{{iOSCloser}}/g, iOSCloser)
}

export async function readFile (file, readDirPath) {
  const fileData = await fs.readFile(path.join(readDirPath, file), 'utf-8')
  return fileData
}

function readAndWriteFiles (files, paths, templateName, packageName = null, app = null) {
  const { readDirPath, writeDirPath } = paths
  return Promise.all(
    files.map(async function(file) {
      const fileData = await readFile(file, readDirPath)
      const parsedFile = parseFile(fileData, templateName, packageName, app)
      const fileName = file.replace('Template', templateName)
      return fs.writeFile(path.join(writeDirPath, fileName), parsedFile)
    })
  )
}

async function createAndroidEnvironment (templateName) {
  const appPath = path.join(
    process.cwd(),
    'android',
    'app',
    'src',
    'main',
    'java',
    'com',
    pkg.name
  )
  const writeDirPath = await mkdir(path.join(appPath, templateName.toLowerCase()))
  const paths = {
    readDirPath: path.join(__dirname, '..', 'templates', 'android'),
    writeDirPath
  }
  const files = ['TemplatePackage.java', 'TemplateModule.java', 'TemplateManager.java']
  return readAndWriteFiles(files, paths, templateName, templateName.toLowerCase(), pkg.name)
}

function createSwiftEnvironment (templateName) {
  const paths = {
    readDirPath: path.join(__dirname, '..', 'templates', 'ios-swift'),
    writeDirPath: path.join(process.cwd(), 'ios')
  }
  const files = ['Template.m', 'TemplateManager.swift']
  return readAndWriteFiles(files, paths, templateName)
}

function createObjCEnvironment (templateName) {
  const paths = {
    readDirPath: path.join(__dirname, '..', 'templates', 'ios-objc'),
    writeDirPath: path.join(process.cwd(), 'ios')
  }
  const files = ['Template.h', 'TemplateManager.m']
  return readAndWriteFiles(files, paths, templateName)
}

async function createJSEnvironment (templateName, jsPath) {
  await mkdir(jsPath)

  const paths = {
    readDirPath: path.join(__dirname, '..', 'templates', 'js'),
    writeDirPath: jsPath
  }
  const files = ['TemplateNativeModule.js', 'TemplateNativeView.js']
  return readAndWriteFiles(files, paths, templateName)
}

init()
