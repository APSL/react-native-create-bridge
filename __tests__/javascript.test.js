import path from 'path'
import { readFile, parseFile } from '../src'

describe('JS', () => {
  const templateName = 'TestModule'
  const readDirPath = path.join(__dirname, '..', 'templates', 'js')

  it('creates a TemplateNativeModule.js', async () => {
    const fileData = await readFile('TemplateNativeModule.js', readDirPath)
    const parsedFile = parseFile(fileData, templateName)
    expect(parsedFile).toMatchSnapshot()
  })

  it('creates a TemplateNativeView.js', async () => {
    const fileData = await readFile('TemplateNativeView.js', readDirPath)
    const parsedFile = parseFile(fileData, templateName)
    expect(parsedFile).toMatchSnapshot()
  })
})
