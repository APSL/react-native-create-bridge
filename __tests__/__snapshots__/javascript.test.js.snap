exports[`JS creates a TemplateNativeModule.js 1`] = `
"//  Created by react-native-create-bridge

import { NativeModules } from \'react-native\'

const { TestModule } = NativeModules

export default {
  exampleMethod () {
    return TestModule.exampleMethod()
  },

  EXAMPLE_CONSTANT: TestModule.EXAMPLE_CONSTANT
}
"
`;

exports[`JS creates a TemplateNativeView.js 1`] = `
"//  Created by react-native-create-bridge

import React, { Component } from \'react\'
import { requireNativeComponent } from \'react-native\'

const TestModule = requireNativeComponent(\'TestModule\', null)

class TestModuleView extends Component {
  render () {
    return <TestModule {...this.props} />
  }
}

TestModuleView.propTypes = {
  exampleProp: React.PropTypes.any
}

export default TestModuleView
"
`;
